### PR TITLE
CLI: Add configure commands

### PR DESCRIPTION
**Pull request by dmabtrg**
This PR adds the `configure`, `configure cygnus`, `configure hosts`, `configure keyrock` and `configure oauth` commands to the CLI. This PR depends on #127.  This PR adds functionality needed by #126.
- `configure cygnus` allows to configure the MySQL user and password for cygnus before starting the service,
- `configure hosts` allows to add the running containers hostname and IP to the `/etc/hosts` file,
- `configure keyrock` allows to provision the sample users, roles and permissions for TourGuide-App,
- `configure oauth` allows to configure the oauth credentials for the TourGuide-App as generated on keyrock after provisioning and reloading this configuration on the tourguide container.

### CODE REVIEW

File: cli/configure_cygnus [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    fi
+}
+
+function submodule_cmd () {
+    submodule_options "$@"
+
+    local sum=$( md5sum "${compose}" )
+    # configure cygnus container
+    sed -i "${compose}" \
+        -e "/^cygnus:/,/^$/ s/CYGNUS_MYSQL_USER=.*$/CYGNUS_MYSQL_USER=${cygnus_mysql_user}/" \
```
Comment (FiwareULPGC) [inline cli/configure_cygnus]: Line is too long.

File: cli/configure_cygnus [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+        -e "/^cygnus:/,/^$/ s/CYGNUS_MYSQL_PASS=.*$/CYGNUS_MYSQL_PASS=${cygnus_mysql_password}/"
```
Comment (FiwareULPGC) [inline cli/configure_cygnus]: Line is too long.

File: cli/configure_cygnus [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+    # configure mysql container
+    if [ "${cygnus_mysql_user}" = "root" ] ; then
+        sed -i "${compose}" \
+            -e "/^mysql:/,/^$/ {/environment:/,/^$/ { /- /d}}" \
+            -e "/^mysql:/,/^$/ {/environment:/ s/environment:.*$/environment:\n        - MYSQL_ROOT_PASSWORD=${cygnus_mysql_password}/}"
```
Comment (FiwareULPGC) [inline cli/configure_cygnus]: Line is too long.

File: cli/configure_cygnus [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    else
+        sed -i "${compose}" \
+            -e "/^mysql:/,/^$/ {/environment:/,/^$/ { /- /d}}" \
+            -e "/^mysql:/,/^$/ {/environment:/ s/environment:.*$/environment:\n        - MYSQL_USER=${cygnus_mysql_user}\n        - MYSQL_PASSWORD=${cygnus_mysql_password}\n        - MYSQL_DATABASE=${fiware_service}/}"
```
Comment (FiwareULPGC) [inline cli/configure_cygnus]: Line is too long.

File: cli/configure_hosts [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+    local user=$( whoami )
+    if [ ${requires_root} -eq 1 ] ; then
+        if [ "${user}" != "root" ] ; then
+            echo "Only root is allowed to modify ${hosts_file}"
+            echo "Run the script as root or use an alternate method like sudo"
+            exit 1
+        fi
+    else
+        if [ ! -w "${hosts_file}" ] ; then
+            echo "Hosts file '${hosts_file}' does not have write permission for user ${user}."
```
Comment (FiwareULPGC) [inline cli/configure_hosts]: Line is too long.

File: cli/configure_hosts [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+    # get list of all containers
+    containers_list=($( docker ps -a -q --no-trunc ))
+    if [ "${#containers_list[@]}" -eq 0 ] ; then
+        echo "No containers found."
+        echo "Hosts file will NOT be modified."
+        exit 0
+    fi
+
+    # get list of running containers
+    containers_running=($( docker inspect --format "{{if .State.Running}}{{.Id}}{{end}}" "${containers_list[@]}" ))
```
Comment (FiwareULPGC) [inline cli/configure_hosts]: Line is too long.

File: cli/configure_hosts [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+    if [ -z "${containers_running}" ] ; then
+        echo "There's no running containers."
+        echo "Hosts file will NOT be modified."
+        exit 0
+    else
+        # get the names of the running containers
+        container_names=( $( docker inspect --format '{{.Name}}' "${containers_running[@]}" ) )
```
Comment (FiwareULPGC) [inline cli/configure_hosts]: Line is too long.

File: cli/configure_keyrock [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+
+    if [ ! -e "${NC}" ] ; then
+        echo "Unable to find 'nc' command."
+        exit 1
+    fi
+
+    echo "Testing if port '${_port}' is open at host '${_host}'."
+
+    while [ ${_tries} -lt ${_max_tries} -a ${_is_open} -eq 0 ] ; do
+        echo -n "Checking connection to '${_host}:${_port}' [try $(( ${_tries} + 1 ))/${_max_tries}] ... "
```
Comment (FiwareULPGC) [inline cli/configure_keyrock]: Line is too long.

File: cli/configure_keyrock [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+    docker cp ${keyrock_scripts_path}/params-config.py ${_container_name}:/opt/
+    docker cp ${keyrock_scripts_path}/access_control_xacml.py ${_container_name}:/horizon/
+
+    # provision tourguide users, roles and permissions
+    echo "Provisioning TourGuide users, roles and permissions."
+    docker exec -i ${_container_name} /bin/bash -c \
+           'cd /keystone ; \
+            source .venv/bin/activate ; \
+            (sleep 5 ; echo idm) | bin/keystone-manage -v db_sync --populate ; \
+            [ ! -e /opt/.provision-done ] && python /opt/tourguide-provision.py && touch /opt/.provision-done'
```
Comment (FiwareULPGC) [inline cli/configure_keyrock]: Line is too long.

File: cli/configure_keyrock [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+
+    # configure access and sync with authzforce
+    echo "Syncing with authzforce."
+    docker exec -i ${_container_name} /bin/bash -c \
+           "sed -i /horizon/openstack_dashboard/local/local_settings.py \
+            -e \"s|^ACCESS_CONTROL_URL = None|ACCESS_CONTROL_URL = '${keyrock_ac_url}'|\" \
+            -e \"s|^ACCESS_CONTROL_MAGIC_KEY = None|ACCESS_CONTROL_MAGIC_KEY = '${keyrock_ac_magic_key}'|\" ; \
```
Comment (FiwareULPGC) [inline cli/configure_keyrock]: Line is too long.

File: cli/configure_keyrock [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+            python /opt/params-config.py --name 'TourGuide' --file /opt/config.json --database keystone.db ; \
```
Comment (FiwareULPGC) [inline cli/configure_keyrock]: Line is too long.

File: cli/configure_oauth [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+
+    if [ ! -e "${NC}" ] ; then
+        echo "Unable to find 'nc' command."
+        exit 1
+    fi
+
+    echo "Testing if port '${_port}' is open at host '${_host}'."
+
+    while [ ${_tries} -lt ${_max_tries} -a ${_is_open} -eq 0 ] ; do
+        echo -n "Checking connection to '${_host}:${_port}' [try $(( ${_tries} + 1 ))/${_max_tries}] ... "
```
Comment (FiwareULPGC) [inline cli/configure_oauth]: Line is too long.

File: cli/configure_oauth [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+            if [ ${_tries} -lt ${_max_tries} ] ; then
+                echo "Retrying."
+            else
+                echo "Failed."
+            fi
+        fi
+    done
+
+    if [ ${_is_open} -eq 0 ] ; then
+        echo "Failed to connect to port '${_port}' on host '${_host}' after ${_tries} tries."
```
Comment (FiwareULPGC) [inline cli/configure_oauth]: Line is too long.

File: cli/configure_oauth [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+    local _started=0
+    local _tries=0
+    local _keyrock_container="${container_prefix}keyrock"
+    local _tourguide_container="${container_prefix}tourguide"
+
+    check_host_port localhost 5000 ${keyrock_wait}
+
+    while [ true ]; do
+        echo -n "Waiting for tourguide to be ready [$(( ${_tries} + 1 ))]... "
+        if ( docker logs ${_tourguide_container} 2>&1 | grep -qE "service apache2 reload" ) ; then
```
Comment (FiwareULPGC) [inline cli/configure_oauth]: Line is too long.

File: cli/configure_oauth [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+            _tries=$(( ${_tries} + 1 ))
+            echo "Retrying."
+        else
+            break
+        fi
+    done
+
+    if [ ${_started} -eq 1 ]; then
+        # get the OAuth credentials for the app
+        CLIENT_ID=$( echo $( docker exec -i ${_keyrock_container} /bin/bash -c 'cat /opt/config.json' ) | grep -Po '(?<="id": ")[^"]*' )
```
Comment (FiwareULPGC) [inline cli/configure_oauth]: Line is too long.

File: cli/configure_oauth [hunk 16]
```diff
@@ -0,0 +1,148 @@
...
+        CLIENT_SECRET=$( echo $( docker exec -i ${_keyrock_container} /bin/bash -c 'cat /opt/config.json' ) | grep -Po '(?<="secret": ")[^"]*' )
```
Comment (FiwareULPGC) [inline cli/configure_oauth]: Line is too long.

File: cli/configure_oauth [hunk 17]
```diff
@@ -0,0 +1,148 @@
...
+        docker exec -i ${_tourguide_container} /bin/bash -c \
+               "sed -i \${CC_APP_SERVER_PATH}/config.js \
+                -e \"s|config.clientId = 'CLIENT_ID'|config.clientId = '${CLIENT_ID}'|g\" \
+                -e \"s|config.clientSecret = 'CLIENT_SECRET'|config.clientSecret = '${CLIENT_SECRET}'|g\" ; \
```
Comment (FiwareULPGC) [inline cli/configure_oauth]: Line is too long.

File: docker/keyrock/tourguide-provision.py [hunk 18]
```diff
@@ -0,0 +1,212 @@
...
+    owner_role = keystone.roles.find(name='owner')
+
+    # Create 10 users
+    users = []
+    for i in range(10):
+        username = 'user'
+        users.append(_register_user(keystone, username + str(i)))
+
+    # Register pepProxy user
+
```
Comment (FiwareULPGC) [inline docker/keyrock/tourguide-provision.py]: Unneeded line.

File: docker/keyrock/tourguide-provision.py [hunk 19]
```diff
@@ -0,0 +1,212 @@
...
+    pep_user = _register_user(keystone, 'pepproxy')
+
+    # Create Franchises
+
```
Comment (FiwareULPGC) [inline docker/keyrock/tourguide-provision.py]: Unneeded line.

File: docker/keyrock/tourguide-provision.py [hunk 20]
```diff
@@ -0,0 +1,212 @@
...
+    franchises = []
+
+    for i in range(4):
+        franchises.append(_create_organization(keystone, 'Franchise' + str(i+1)))
+
+
```
Comment (FiwareULPGC) [inline docker/keyrock/tourguide-provision.py]: Unneeded line.

File: docker/keyrock/tourguide-provision.py [hunk 21]
```diff
@@ -0,0 +1,212 @@
...
+        is_internal=False,
+        application=tourguide_app.id)
+
+    # Global manager
+    global_manager = keystone.fiware_roles.roles.create(
+        name='Global manager',
+        is_internal=False,
+        application=tourguide_app.id)
+
+
```
Comment (FiwareULPGC) [inline docker/keyrock/tourguide-provision.py]: Unneeded line.

File: docker/keyrock/tourguide-provision.py [hunk 22]
```diff
@@ -0,0 +1,212 @@
...
+    # Make all users Restaurant viewers
+
```
Comment (FiwareULPGC) [inline docker/keyrock/tourguide-provision.py]: Unneeded line.

File: docker/keyrock/tourguide-provision.py [hunk 23]
```diff
@@ -0,0 +1,212 @@
...
+    for user in users:
+        keystone.fiware_roles.roles.add_to_user(
+            role=end_user.id,
+            user=user.id,
+            application=tourguide_app.id,
+            organization=user.default_project_id)
+
+    # Make user0 Global Manager
+
```
Comment (FiwareULPGC) [inline docker/keyrock/tourguide-provision.py]: Unneeded line.

File: docker/keyrock/tourguide-provision.py [hunk 24]
```diff
@@ -0,0 +1,212 @@
...
+        user=users[4].id,
+        application=tourguide_app.id,
+        organization=franchises[3].id)
+
+    for i in range(4):
+        keystone.roles.grant(user=users[i+1].id,
+            role=owner_role.id,
+            project=franchises[i].id)
+
+
```
Comment (FiwareULPGC) [inline docker/keyrock/tourguide-provision.py]: Unneeded line.

File: docker/keyrock/tourguide-provision.py [hunk 25]
```diff
@@ -0,0 +1,212 @@
...
+    # Make user1-4 Frnanchise Manager
+
+    # Adding permissions for manager and restaurants (TODO)
+
```
Comment (FiwareULPGC) [inline docker/keyrock/tourguide-provision.py]: Unneeded line.

**Comment by FiwareULPGC** [comment]
`./touguide configure oauth` does not recognize the `--help`  option.
oauth-no-help

**Comment by dmabtrg** [comment]
> `./touguide configure oauth` does not recognize the `--help`  option.

Fixed.

File: cli/configure_cygnus (hunk 1 above)
Comment (dmabtrg) [reply cli/configure_cygnus]: Fixed.

File: cli/configure_cygnus (hunk 2 above)
Comment (dmabtrg) [reply cli/configure_cygnus]: Fixed.

File: cli/configure_cygnus (hunk 3 above)
Comment (dmabtrg) [reply cli/configure_cygnus]: Fixed.

File: cli/configure_cygnus (hunk 4 above)
Comment (dmabtrg) [reply cli/configure_cygnus]: Fixed.

File: cli/configure_hosts (hunk 5 above)
Comment (dmabtrg) [reply cli/configure_hosts]: Fixed.

File: cli/configure_hosts (hunk 6 above)
Comment (dmabtrg) [reply cli/configure_hosts]: Fixed.

File: cli/configure_hosts (hunk 7 above)
Comment (dmabtrg) [reply cli/configure_hosts]: Fixed.

File: cli/configure_keyrock (hunk 8 above)
Comment (dmabtrg) [reply cli/configure_keyrock]: Fixed.

File: cli/configure_keyrock (hunk 9 above)
Comment (dmabtrg) [reply cli/configure_keyrock]: Fixed.

File: cli/configure_keyrock (hunk 10 above)
Comment (dmabtrg) [reply cli/configure_keyrock]: Fixed.

File: cli/configure_keyrock (hunk 11 above)
Comment (dmabtrg) [reply cli/configure_keyrock]: Fixed.

File: cli/configure_oauth (hunk 12 above)
Comment (dmabtrg) [reply cli/configure_oauth]: Fixed.

File: cli/configure_oauth (hunk 13 above)
Comment (dmabtrg) [reply cli/configure_oauth]: Fixed.

File: cli/configure_oauth (hunk 14 above)
Comment (dmabtrg) [reply cli/configure_oauth]: Fixed.

File: cli/configure_oauth (hunk 15 above)
Comment (dmabtrg) [reply cli/configure_oauth]: Fixed.

File: cli/configure_oauth (hunk 16 above)
Comment (dmabtrg) [reply cli/configure_oauth]: Fixed.

File: cli/configure_oauth (hunk 17 above)
Comment (dmabtrg) [reply cli/configure_oauth]: Fixed.

File: docker/keyrock/tourguide-provision.py (hunk 18 above)
Comment (dmabtrg) [reply docker/keyrock/tourguide-provision.py]: Fixed.

File: docker/keyrock/tourguide-provision.py (hunk 19 above)
Comment (dmabtrg) [reply docker/keyrock/tourguide-provision.py]: Fixed.

File: docker/keyrock/tourguide-provision.py (hunk 20 above)
Comment (dmabtrg) [reply docker/keyrock/tourguide-provision.py]: Fixed.

File: docker/keyrock/tourguide-provision.py (hunk 21 above)
Comment (dmabtrg) [reply docker/keyrock/tourguide-provision.py]: Fixed.

File: docker/keyrock/tourguide-provision.py (hunk 22 above)
Comment (dmabtrg) [reply docker/keyrock/tourguide-provision.py]: Fixed.

File: docker/keyrock/tourguide-provision.py (hunk 23 above)
Comment (dmabtrg) [reply docker/keyrock/tourguide-provision.py]: Fixed.

File: docker/keyrock/tourguide-provision.py (hunk 24 above)
Comment (dmabtrg) [reply docker/keyrock/tourguide-provision.py]: Fixed.

File: docker/keyrock/tourguide-provision.py (hunk 25 above)
Comment (dmabtrg) [reply docker/keyrock/tourguide-provision.py]: Fixed.

**Comment by FiwareULPGC** [comment]
LGTM

**Pull request merged**